ydb/core/tx/schemeshard/schemeshard__login.cpp

```diff
@@ -67,37 +69,131 @@ struct TSchemeShard::TTxLogin : TSchemeShard::TRwTxBase {
 
             NeedPublishOnComplete = true;
         }
+
+        return LoginAttempt(db, ctx);
     }
 
-    void DoComplete(const TActorContext &ctx) override {
+    void Complete(const TActorContext &ctx) override {
         if (NeedPublishOnComplete) {
             Self->PublishToSchemeBoard(TTxId(), {SubDomainPathId}, ctx);
         }
 
-        THolder<TEvSchemeShard::TEvLoginResult> result = MakeHolder<TEvSchemeShard::TEvLoginResult>();
+        LOG_DEBUG_S(ctx, NKikimrServices::FLAT_TX_SCHEMESHARD,
+                    "TTxLogin Complete"
+                    << ", result: " << Result->Record.ShortDebugString()
+                    << ", at schemeshard: " << Self->TabletID());
+
+        ctx.Send(Request->Sender, std::move(Result), 0, Request->Cookie);
+    }
+
+private:
+    bool LoginAttempt(NIceDb::TNiceDb& db, const TActorContext& ctx) {
         const auto& loginRequest = GetLoginRequest();
-        if (loginRequest.ExternalAuth || AppData(ctx)->AuthConfig.GetEnableLoginAuthentication()) {
-            NLogin::TLoginProvider::TLoginUserResponse loginResponse = Self->LoginProvider.LoginUser(loginRequest);
-            if (loginResponse.Error) {
-                result->Record.SetError(loginResponse.Error);
-            }
-            if (loginResponse.Token) {
-                result->Record.SetToken(loginResponse.Token);
-                result->Record.SetSanitizedToken(loginResponse.SanitizedToken);
+        if (!loginRequest.ExternalAuth && !AppData(ctx)->AuthConfig.GetEnableLoginAuthentication()) {
+            Result->Record.SetError("Login authentication is disabled");
+            return true;
+        }
+        if (loginRequest.ExternalAuth) {
+            return HandleExternalAuth(loginRequest);
+        }
+        return HandleLoginAuth(loginRequest, db, ctx);
+    }
+
+    bool HandleExternalAuth(const NLogin::TLoginProvider::TLoginUserRequest& loginRequest) {
+        const NLogin::TLoginProvider::TLoginUserResponse loginResponse = Self->LoginProvider.LoginUser(loginRequest);
+        switch (loginResponse.Status) {
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::SUCCESS: {
+            SetSuccessResult(loginResponse);
+            break;
+        }
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::INVALID_PASSWORD:
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::INVALID_USER:
+        case NLogin::TLoginProvider::TLoginUserResponse::EStatus::UNAVAILABLE_KEY: {
```

Comment:
Is it covered with unit tests?

Reply:
Cannot test code returns this error. I think this piece of code is unreachable